src/app/shared/components/batchupload/batchupload.component.ts

```diff
@@ -0,0 +1,370 @@
+    });
+    this._settingsservice.getEntities(this.configSettings.nssBaseURL + this.configSettings.unitsURL).subscribe((unitTypes: Array<Unittype>) => {
+      this.unitTypes = unitTypes;
+    });
+    this._settingsservice.getEntities(this.configSettings.nssBaseURL + this.configSettings.regTypeURL).subscribe((regTypes: Array<Regressiontype>)  => {
+      this.regressionTypes = regTypes;
+    });
+    this._settingsservice.getEntities(this.configSettings.nssBaseURL + this.configSettings.errorsURL).subscribe((errorTypes: Array<Error>) => {
+      this.errors = errorTypes;
+    });
+  }
+
+  public showModal(): void {
+    this.modalRef = this._modalService.open(this.modalElement, { backdrop: 'static', keyboard: false, size: 'lg', windowClass: 'modal-xl batch-upload-modal' });
+  }
+
+  public clearTable() {
+    this.sheetNamesButtons = false;
+    this.tableDisplay = false; 
+    delete(this.data);
+  }
+
+  public selectFile(event: any) {
+    const target: DataTransfer = <DataTransfer>(event.target);
+    if (target.files.length !== 1) { //Check for multiple files
+      this.clearTable();
+      this._toasterService.pop('error', 'Error', 'Cannot select multiple files');
+      return;
+    } 
+    var ext = event.target.files[0].name.match(/\.([^\.]+)$/)[1];
+    switch (ext) {  //check for incompatible file type
+      case 'xlsx':
+      case 'xls':
+        break;
+      default:
+        this.clearTable();
+        this._toasterService.pop('error', 'Error', 'File type must be .xlsx');
+        return;
+    }
+    const reader: FileReader = new FileReader();
+    reader.onload = (e:any) => {
+        const bstr: string = e.target.result;                          
+        this.wb = XLSX.read(bstr, { type: 'binary'});          // Read WorkBook
+        this.sheetNamesButtons = true;                         // Show buttons to select worksheet from workbook
+        };
+    reader.readAsBinaryString(target.files[0]); 
+  }
+
+  public selectSheet(sheetName) {
+    const ws: XLSX.WorkSheet = this.wb.Sheets[sheetName];
+    this.data = (XLSX.utils.sheet_to_json(ws, {header : 1}));        // Convert data to json
+    this.createTable(this.data);
+    this.tableDisplay = true;
+    this.sheetNamesButtons = false;
+  }
+
+  async createTable(data) {
+    this.success = [];
+    this.clicked = false;
+    var counter = 0;
+    var explanatoryVariablesArray: explanatoryVariables[] = [];
+    var tempExplanatoryVariablesArray: explanatoryVariables[] = [];
+    var errorsArray:errors[] = [];
+    var explanatoryVariables: explanatoryVariables = {  
+      code: null,
+      limits: ({
+          max: null,
+          min: null,
+      }),
+      unitType: { 
+        id: null 
+      }
+    };
+    var errors: errors = {  
+      id: null,
+      value:  null 
+    };
+
+    // Loop through spreadsheet
+    for (var i = 2; i < data.length; i++) { 
+      if (data[i][0]) { // Study area
+        var studyArea = (data[i][0]);
```

Comment:
If I'm understanding correctly, I think it would make more sense to initialize these types of variables (`var studyArea`) above the loop. It seems like the loop is using whatever was last set as the studyArea, statistic group, etc. (since not every line will have them). If that's the case, we can initiate this variable above so anyone who works on this in the future is clear that this is a variable that is used in every loop, whether or not it's set in that current iteration. Hope this makes sense 😬